File: 107. Binary Tree Level Order Traversal II
/binaryTreeLevelOrderTraversalII.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque
class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        result = []
        if root == None:
            return result
        q = deque()
        q.append(root)
        while q:
            level = []
            levelSize = len(q)
            for _ in range(levelSize):
                node = q.pop()
                if node.left:
                    q.appendleft(node.left)
                if node.right:
                    q.appendleft(node.right)
                level.append(int(node.val))
            result.append(level)
        return reversed(result)

File: 107. Binary Tree Level Order Traversal II
/binaryTreeLevelOrderTraversalII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <queue>
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> result;
        if(root == NULL) return result;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            vector<int> level;
            int levelSize = q.size();
            for(int i = 0; i < levelSize; i++){
                auto node = q.front();
                q.pop();
                level.push_back(node->val);
                if(node->left != NULL) q.push(node->left);
                if(node->right != NULL) q.push(node->right);   
            }
            result.push_back(level);
        }
        reverse(result.begin(), result.end());
        return result;
    }
    
};
